lib/DependencyScan/ModuleDependencyScanner.cpp

```diff
@@ -1473,33 +1466,26 @@ void ModuleDependencyScanner::resolveHeaderDependenciesForModule(
             moduleDependencyInfo.updateBridgingHeaderCommandLine(
                 bridgingHeaderCommandLine);
           moduleDependencyInfo.setHeaderSourceFiles(headerFileInputs);
+          // Update the set of visible Clang modules
+          moduleDependencyInfo.addVisibleClangModules(visibleClangModules);
           // Update the dependency in the cache
           cache.updateDependency(moduleID, moduleDependencyInfo);
         } else {
           // Failure to scan header
         }
         return true;
       });
-  cache.setHeaderClangDependencies(moduleID,
-                                   headerClangModuleDependencies.getArrayRef());
 }
 
 void ModuleDependencyScanner::resolveSwiftOverlayDependenciesForModule(
     const ModuleDependencyID &moduleID, ModuleDependenciesCache &cache,
     ModuleDependencyIDSetVector &swiftOverlayDependencies) {
   PrettyStackTraceStringAction trace(
       "Resolving Swift Overlay dependencies of module", moduleID.ModuleName);
-  std::vector<std::string> allClangDependencies;
-  llvm::StringSet<> knownModules;
-
-  // Find all of the discovered Clang modules that this module depends on.
-  for (const auto &dep : cache.getClangDependencies(moduleID))
-    findAllImportedClangModules(dep.ModuleName, cache, allClangDependencies,
-                                knownModules);
-
+  auto visibleClangDependencies = cache.getVisibleClangModules(moduleID);
```

Comment:
question regarding how the cache works with https://github.com/swiftlang/swift/blob/15593e5ca68b7e3fba358fe617fd38fc66bdf879/lib/DependencyScan/ModuleDependencyScanner.cpp#L1127 
and https://github.com/swiftlang/swift/blob/15593e5ca68b7e3fba358fe617fd38fc66bdf879/lib/DependencyScan/ModuleDependencyScanner.cpp#L1176

Consider a direct dependency clang module `Foo` that has a transitive dependency `Bar`. If I read this correctly, 
after getting `Foo`'s dependency info from clang, there would be a cache hit on `Bar`'s moduleID if it was later observed as a direct dependency. But would there be information about the visible modules from importing `Bar`? I suspect the correct behavior would be to ask the clang scanning service.

Reply:
That's a great catch!
Yeah, since the visible module set is per-query and not per-module, I think we might have to query every single module directly imported from Swift, even if it's been previously reported as a transitive dependency. I'm gonna give some thought to how best to organize that and see if I can measure what perf impact this will have.